internal/batches/service/service.go

```diff
@@ -157,20 +166,74 @@ func (svc *Service) CreateChangesetSpec(ctx context.Context, spec *batcheslib.Ch
 // Progress information is reported back to the given progress function: perc
 // will be a value between 0.0 and 1.0, inclusive.
 func (svc *Service) EnsureDockerImages(ctx context.Context, spec *batcheslib.BatchSpec, progress func(done, total int)) (map[string]docker.Image, error) {
-	total := len(spec.Steps)
+	// Figure out the image names used in the batch spec.
+	names := map[string]struct{}{}
+	for i := range spec.Steps {
+		names[spec.Steps[i].Container] = struct{}{}
+	}
+
+	// Set up our progress reporting.
+	total := len(names)
 	progress(0, total)
 
-	// TODO: this _really_ should be parallelised, since the image cache takes
-	// care to only pull the same image once.
+	// Set up the channels that will be used in the parallel goroutines handling
+	// the pulls.
+	type image struct {
+		name  string
+		image docker.Image
+		err   error
+	}
+	complete := make(chan image)
+	inputs := make(chan string, total)
+
+	// Set up a worker context that we can use to terminate the workers if an
+	// error occurs.
+	workerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Spawn worker goroutines to call EnsureImage on each image name.
+	var wg sync.WaitGroup
+	for i := 0; i < svc.parallelism; i++ {
+		wg.Add(1)
+		go func() {
+			for name := range inputs {
+				img, err := svc.EnsureImage(workerCtx, name)
+				complete <- image{
+					name:  name,
+					image: img,
+					err:   err,
+				}
+			}
+			wg.Done()
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(complete)
+	}()
+
+	// Send the image names to the worker goroutines.
+	go func() {
+		for name := range names {
+			inputs <- name
+		}
+		close(inputs)
+	}()
+
+	// Receive the results of the image pulls and build the return value.
+	i := 0
 	images := make(map[string]docker.Image)
-	for i := range spec.Steps {
-		img, err := svc.EnsureImage(ctx, spec.Steps[i].Container)
-		if err != nil {
-			return nil, err
+	for image := range complete {
+		if image.err != nil {
+			// If EnsureImage errored, then we'll early return here and let the
+			// worker context clean things up.
+			return nil, image.err
```

Comment:
Does that work? The goroutines don't check for `ctx.Done()` and could be stucking waiting to put something in `complete` (which is unbuffered) because there's no reader on the other side. And the goroutine that does `wg.Wait()` would be stuck on that.

Example: the `return` here fires but another goroutine just returned from `EnsureImage` above and now tries to put something in `complete`. It will get stuck.

I think you need something like

```go
select {
case <-ctx.Done():
  return
case  complete <- image{name:  name, image: img,err:   err }:
  // all good
}
```

in your goroutine above.

(And possibly another check for `ctx.Done()` before `EnsureImage` so we don't download images if one failed)

Reply:
Fixed in 180078d (and tested in 188a328).